cmd/prometheus/main.go

```diff
@@ -472,7 +472,9 @@ func main() {
 
 			},
 			func(err error) {
-				close(cancel)
+				// Wait for any in-progress reloads to complete to avoid
+				// reloading things after they have been shutdown.
+				cancel <- struct{}{}
```

Comment:
I don't quite understand what this changes. This will make this interrupt function block until the above function actually reached the point where it reads from `cancel`. But what benefit does that give us?

Reply:
It ensures that there will never be a reload after that read.

Reply:
I don't follow. Once the above function receives something from cancel it returns immediately.
Whether that happens through closing `cancel` or sending a value on it doesn't matter.

Reply:
It's an unbuffered channel, so if there's a reload in progress it'll block.

Reply:
Interrupt functions are called sequentially, without concurrency.

Reply:
That's an implementation detail, and I'd think @peterbourgon would object that this is something to rely on. Maybe he can provide more context here.

https://godoc.org/github.com/oklog/run#Group.Run

Reply:
I take this is also the reason why you changed the order in which `Group.Add`s are called?

Reply:
Yes, to control the shutdown order.

Reply:
I also think that we shouldn't depend on the implementation details of `Group.Run()` and same for the ordering of `Group.Add()`. That's too implicit to me and easy to screw up when not paying close attention.

It'd be better to impose those ordering constraints explicitly externally, like we do with some of the other ordering channels already.

Reply:
I was under the impression that the whole point of switching to this library was to get rid of the ad-hoc approaches we used previously. Are you proposing we get rid of this library and go back to doing all this ourselves?

Reply:
Maybe it's a good opportunity for someone to create a library that allows more explicit modeling of ordering constraints (half serious) :)

I like the library usage here already, but when it comes to ordering, it just doesn't give you any tools (besides implementation details), so we do still have to do that ourselves.

Reply:
Reading the docs more closely, this is the intended behaviour. The first line is "Package run implements an actor-runner with *deterministic teardown*", which you can only offer if the interrupts are sequential and synchronous.

Reply:
Before I get into detail, one clarification: it looks like you're trying to prevent subsequent components, like ruleManager, from _being interrupted_ until the reloadHandler's execute func has returned. Is that correct?

Reply:
Basically yes. The rulemanager shouldn't shutdown until the reload handler has.

Reply:
OK, thanks. This is a super gray area.

First, I didn't mean _deterministic_ as in _in a specific order_ but rather _in any order, but guaranteed_ — I'll push a doc update to make that more clear. So you are indeed relying on implementation details at the moment. This implementation detail is the only sane way to accomplish the work, so it could feasibly be lifted into the contract of the API. But I'm hesitant to do that, because of situations precisely like this one.

From my admittedly rather strict perspective, if ruleManager shouldn't terminate until reloadHandler has terminated, then that causal dependency should be expressed explicitly in the components themselves. A strawman for argument's sake:

```go
type ruleManager struct {
    reloadHandler interface{ Done() <-chan struct{} } // or whatever
    // ...
}

func (rm *ruleManager) Run() {
    defer func() {
        <-rm.reloadHandler.Done()
    }()
    // ...
}
```

I think this is strictly superior, especially for future maintainers, who now understand the required causality rules explicitly, without having to build a mental model that relies on e.g. source code order and guessing at the behavior of the run.Group black box.

But I say it's a gray area, because sometimes the practicalities of building software outweigh a theoretical soundness. If you were whipping up a prototype and wanted to ensure a certain startup/shutdown order, maybe it's fine to do that by leveraging details of your component lifecycle manager package. Maybe that gets the job done reliably, and you can move on to the next thing.

Which kind of brings me full-circle, actually. If you're writing that prototype, then implementation details are a fine-enough "contract" for you to leverage in your code. And if you're writing a more substantial project with lots of maintainers, then I don't want to encourage you to encode dependency relationships via the order of calling functions on an opaque type. It's certainly better to lift these important details of your implementation into the implementation itself, and make them explicit.

With all of this typed out, I'm open to hearing the counterpoints.

Reply:
> then that causal dependency should be expressed explicitly in the components themselves.

That'd be a layering violation, these components don't know that each other exist - nor should they. All the glue code is meant to live inside this file, and I'm trying to enforce that reloads don't happen after shutdown has commenced.

Reply:
> That'd be a layering violation, these components don't know that each other exist - nor should they. All the glue code is meant to live inside this file, and I'm trying to enforce that reloads don't happen after shutdown has commenced.

OK, that makes sense to me. I think the current state of things, while a compromise in the ways I've laid out above, is about as good as it gets. I can't imagine a goroutine lifecycle manager with more features that would be a net benefit in clarity, and I think what you have now is clearly superior to rolling it all by hand.

Reply:
IMO the layering violation is less bad because it's very explicit and visible to any new (or old) developer. We already do that for other concerns in this group of actors as well (like `reloadReady` and `dbOpen`), so it's an established pattern.

Reply:
reloadReady and dbOpen exist entirely in this file, not in the object they're managing.

Reply:
I looked at the problem only briefly but I suspect that the main problem comes from how we handle the config management and reloading.  These kind of issues happened few times already and although the components are very well separated still depend on a global config state.
Last time i looked had the impression that it could be solved by introducing another config  layer/manager

Reply:
> reloadReady and dbOpen exist entirely in this file, not in the object they're managing.

I'm not sure I understand how this would mean introducing anything outside of this file? I was thinking of e.g. a channel that gets closed when the config reloader has exited so that others only shut down when this has occurred. But now actually if termination handlers are executed sequentially, this wouldn't help us because it'd either be superfluous or lead to a deadlock if the ordering of groups is wrong.

So that seems like a good argument why the external channel-based coordination only works for startup, not shutdown, because actors are started in parallel, but shut down sequentially.

If we go for the group ordering solution here, it would be great to also have a big fat comment at the top of each actor that explains ordering requirements though.

Reply:
Question: we want TSDB to be started first and stopped after everything else terminated. We also want the first config load to complete successfully. Can we just pull those out of `run.Group` in regular `main` and only use the group for things that are actually independent?

Reply:
I think we're getting a bit sidetracked here. I propose we get this in now as an immediate fix so we can move towards ending the moratorium, and changes beyond that can be done later.

Reply:
@brian-brazil Ok.

Reply:
Is there anyone who objects to merging this as-is for now?